### DISCUSSION TITLE
Integration with HiveMQ

### DISCUSSION THREAD

**Discussion by lonejack** [cesanta/mongoose#2615] [Ask questions, get answers] [answered]
Good morning,
I'm trying to do some experiment with hivemq but at moment I've no results...
Let me describe the problem:
Created account under hivemq for ServerLess account:
![image](https://github.com/cesanta/mongoose/assets/933876/f24f0f54-f14b-4af9-a2bf-666b21dd580d)

Created two user with all privileges "#":
![image](https://github.com/cesanta/mongoose/assets/933876/069170bb-19b4-44f2-a4f5-07723b57488f)

Then, checkout of mongoose project and opened the example "mongoose/examples/mqtt-client". Then applied my data (main.c):
```
//    4. Receives that message back from the subscribed topic and closes
//    5. Timer-based reconnection logic revives the connection when it is down
//
// To enable SSL/TLS, see https://mongoose.ws/tutorials/tls/#how-to-build

#include "mongoose.h"

static const char *s_url =
    "mqtts://abd356af42190fc234d7a901bcf21345.s1.eu.hivemq.cloud:8883/mqtt";
static const char *s_sub_topic = "mg/+/test";     // Publish topic
static const char *s_pub_topic = "mg/clnt/test";  // Subscribe topic
static const char *s_user = "lonejack";         // User
static const char *s_pass = "T5%w8918293sd";  // Password
static int s_qos = 1;                           // MQTT QoS
static struct mg_connection *s_conn;            // Client connection `

static void fn(struct mg_connection *c, int ev, void *ev_data) {
  if (ev == MG_EV_OPEN) {
    MG_INFO(("%lu CREATED", c->id));
    // c->is_hexdumping = 1;
  } else if (ev == MG_EV_CONNECT) {
    if (mg_url_is_ssl(s_url)) {
      struct mg_tls_opts opts = {
          .ca = mg_unpacked("isrgrootx1.pem"),
..

// on timer fn added user and password:

// Timer function - recreate client connection if it is closed
static void timer_fn(void *arg) {
  struct mg_mgr *mgr = (struct mg_mgr *) arg;
  struct mg_mqtt_opts opts = {.clean = true,
                              .qos = s_qos,
                              .topic = mg_str(s_pub_topic),
                              .user = mg_str(s_user),
                              .pass = mg_str(s_pass),
                              .version = 4,
                              .message = mg_str("bye")};
  if (s_conn == NULL) s_conn = mg_mqtt_connect(mgr, s_url, &opts, fn, NULL);
}
```
Instructions inserted or changed respect to original example (the data here inserted is just for the current example):

static const char *s_url = "**mqtts://abd356af42190fc234d7a901bcf21345.s1.eu.hivemq.cloud:8883\/mqtt\\**"; 
static const char *s_user = "**lonejack**";         // User
static const char *s_pass = "**T5%w8918293sd**";  // Password
And the CA
.ca = mg_unpacked("**isrgrootx1.pem**")

// Timer function - recreate client connection if it is closed
static void timer_fn(void *arg) {
                              ...
                              **.user = mg_str(s_user),**
                              **.pass = mg_str(s_pass),**  

Then compiled as indicated:
make CFLAGS_EXTRA="-DMG_TLS=MG_TLS_MBED -lmbedtls -lmbedcrypto -lmbedx509"
This is the output:
```
claudio@claudio-i7:~/extension/stm32/mongoose/examples/mqtt-client$ make CFLAGS_EXTRA="-DMG_TLS=MG_TLS_MBED -lmbedtls -lmbedcrypto -lmbedx509"
./example                    
58c340 2 main.c:35:fn                   1 CREATED
58c3e9 2 main.c:52:fn                   1 CONNECTED to mqtts://f4da62c516534ac98c18588f0646abdb.s1.eu.hivemq.cloud:8883/mqtt
58c3e9 2 main.c:58:fn                   1 SUBSCRIBED to mg/+/test
58c3e9 2 main.c:65:fn                   1 PUBLISHED hello -> mg/clnt/test
58c3e9 1 mqtt.c:450:mqtt_cb             1 MQTT auth failed, code 5
58c3e9 2 main.c:73:fn                   1 CLOSED
58cfa4 2 main.c:35:fn                   3 CREATED
58d056 2 main.c:52:fn                   3 CONNECTED to mqtts://f4da62c516534ac98c18588f0646abdb.s1.eu.hivemq.cloud:8883/mqtt
58d056 2 main.c:58:fn                   3 SUBSCRIBED to mg/+/test
58d056 2 main.c:65:fn                   3 PUBLISHED hello -> mg/clnt/test
58d056 1 mqtt.c:450:mqtt_cb             3 MQTT auth failed, code 5
58d056 2 main.c:73:fn                   3 CLOSED
^C
```
The connection don't remain up. I get always MQTT auth failed.
Can you help me to understand where is my mistake?
Thank you

**Comment by scaprile** [comment, marked as answer]
Please remove your password from this post
MQTT auth failed, code 5: http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718035 (3.2.2.3) -> Not authorized
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718042

Looks like either your password is wrong or your plan does not allow you to publish. Did you even try with mosquitto_pub or any other computer client ?

**Reply by lonejack** [reply]
The password is random (not real) like the server prefix.

_Looks like either your password is wrong or your plan does not allow you to publish. Did you even try with mosquitto_pub or any other computer client ?_ (I'll do today).

So you confirm that (in general) the setup is correct...

**Reply by scaprile** [reply]
can't do a deep check on your code, that is not for us to do. Our examples work, if you followed one of them, then things should work.
I don't see anything that catches my eye.
I see that you connect, you subscribe, you publish, and then you get an error.
I then assume that TCP connected, TLS did the handshake so your certs should be good, so you managed to pack them, MQTT connected. Then, I have to assume the problem is with that error code.

If you need more help, I need to see the traffic, try without TLS or use a verbose computer client

**Reply by lonejack** [reply]
If I run your version works fine:
![image](https://github.com/cesanta/mongoose/assets/933876/68e555f7-2e26-4e22-b39a-5d02c689e4ed)

**Reply by scaprile** [reply]
We don't have a plan.
We don't use user and password.
We didn't read login requirements and authorization levels because we don't use them
Again: 
> If you need more help, I need to see the traffic, try without TLS or use a verbose computer client

(The no TLS suggestion is so you're able to capture your traffic with Wireshark)

**Reply by lonejack** [reply]
Hi Sergio, does the mongoose handle the [Server Name Indication](https://en.wikipedia.org/wiki/Server_Name_Indication)?
I found a nice post about hivemq:
[Need certificate to achieve SSL connexion](https://community.hivemq.com/t/need-certificate-to-achieve-ssl-connexion/1387)

This feature seems handled [Since 7.10](https://github.com/cesanta/mongoose/releases/tag/7.10).

[And Mbed-Tsl handle](https://mbed-tls.readthedocs.io/en/latest/kb/how-to/use-sni/)

**Reply by lonejack** [reply]
Found the problem..
![image](https://github.com/cesanta/mongoose/assets/933876/910239f2-01da-4233-8eba-f2860b831090)
I'm compiling under examples/mqtt-client.
Simply I changed this:
```
struct mg_tls_opts opts = {
          .ca = mg_unpacked("/home/claudio/extension/stm32/nucleo-h755zi-demo/"
                            "extra/isrgrootx1.pem"),
```
with this:
```
      struct mg_tls_opts opts = {.ca = mg_str(s_ca_cert),
                                 .name = mg_url_host(s_url)};
```

And used the certificate indicated by hivemq (by string):
```
static const char *s_ca_cert =
    "-----BEGIN CERTIFICATE-----\n"
    "MIIFFjCCAv6gAwIBAgIRAJErCErPDBinU/bWLiWnX1owDQYJKoZIhvcNAQELBQAw\n"
    "TzELMAkGA1UEBhMCVVMxKTAnBgNVBAoTIEludGVybmV0IFNlY3VyaXR5IFJlc2Vh\n"
    "cmNoIEdyb3VwMRUwEwYDVQQDEwxJU1JHIFJvb3QgWDEwHhcNMjAwOTA0MDAwMDAw\n"
    "WhcNMjUwOTE1MTYwMDAwWjAyMQswCQYDVQQGEwJVUzEWMBQGA1UEChMNTGV0J3Mg\n"
    "RW5jcnlwdDELMAkGA1UEAxMCUjMwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEK\n"
    "AoIBAQC7AhUozPaglNMPEuyNVZLD+ILxmaZ6QoinXSaqtSu5xUyxr45r+XXIo9cP\n"
    "R5QUVTVXjJ6oojkZ9YI8QqlObvU7wy7bjcCwXPNZOOftz2nwWgsbvsCUJCWH+jdx\n"
    "sxPnHKzhm+/b5DtFUkWWqcFTzjTIUu61ru2P3mBw4qVUq7ZtDpelQDRrK9O8Zutm\n"
    "NHz6a4uPVymZ+DAXXbpyb/uBxa3Shlg9F8fnCbvxK/eG3MHacV3URuPMrSXBiLxg\n"
    "Z3Vms/EY96Jc5lP/Ooi2R6X/ExjqmAl3P51T+c8B5fWmcBcUr2Ok/5mzk53cU6cG\n"
    "/kiFHaFpriV1uxPMUgP17VGhi9sVAgMBAAGjggEIMIIBBDAOBgNVHQ8BAf8EBAMC\n"
    "AYYwHQYDVR0lBBYwFAYIKwYBBQUHAwIGCCsGAQUFBwMBMBIGA1UdEwEB/wQIMAYB\n"
    "Af8CAQAwHQYDVR0OBBYEFBQusxe3WFbLrlAJQOYfr52LFMLGMB8GA1UdIwQYMBaA\n"
    "FHm0WeZ7tuXkAXOACIjIGlj26ZtuMDIGCCsGAQUFBwEBBCYwJDAiBggrBgEFBQcw\n"
    "AoYWaHR0cDovL3gxLmkubGVuY3Iub3JnLzAnBgNVHR8EIDAeMBygGqAYhhZodHRw\n"
    "Oi8veDEuYy5sZW5jci5vcmcvMCIGA1UdIAQbMBkwCAYGZ4EMAQIBMA0GCysGAQQB\n"
    "gt8TAQEBMA0GCSqGSIb3DQEBCwUAA4ICAQCFyk5HPqP3hUSFvNVneLKYY611TR6W\n"
    "PTNlclQtgaDqw+34IL9fzLdwALduO/ZelN7kIJ+m74uyA+eitRY8kc607TkC53wl\n"
    "ikfmZW4/RvTZ8M6UK+5UzhK8jCdLuMGYL6KvzXGRSgi3yLgjewQtCPkIVz6D2QQz\n"
    "CkcheAmCJ8MqyJu5zlzyZMjAvnnAT45tRAxekrsu94sQ4egdRCnbWSDtY7kh+BIm\n"
    "lJNXoB1lBMEKIq4QDUOXoRgffuDghje1WrG9ML+Hbisq/yFOGwXD9RiX8F6sw6W4\n"
    "avAuvDszue5L3sz85K+EC4Y/wFVDNvZo4TYXao6Z0f+lQKc0t8DQYzk1OXVu8rp2\n"
    "yJMC6alLbBfODALZvYH7n7do1AZls4I9d1P4jnkDrQoxB3UqQ9hVl3LEKQ73xF1O\n"
    "yK5GhDDX8oVfGKF5u+decIsH4YaTw7mP3GFxJSqv3+0lUFJoi5Lc5da149p90Ids\n"
    "hCExroL1+7mryIkXPeFM5TgO9r0rvZaBFOvV2z0gp35Z0+L4WPlbuEjN/lxPFin+\n"
    "HlUjr8gRsI3qfJOQFy/9rKIJR0Y/8Omwt/8oTWgy1mdeHmmjk7j1nYsvC9JSQ6Zv\n"
    "MldlTTKB3zhThV1+XWYp6rjd5JW1zbVWEkLNxE7GJThEUG3szgBVGP7pSWTUTsqX\n"
    "nLRbwHOoq7hHwg==\n"
    "-----END CERTIFICATE-----\n";
```
![image](https://github.com/cesanta/mongoose/assets/933876/7e2020dd-c72d-4ead-be42-0585924d8687)

**Reply by scaprile** [reply]
Please follow the guidelines in the [documentation](https://mongoose.ws/documentation/),  [examples](https://github.com/cesanta/mongoose/tree/master/examples/) and [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples) available.

https://mongoose.ws/documentation/tutorials/tls/#tls-for-clients
> Optionally, you may also set name to enable host name validation. If you also omit the ca entry, no certificate verification will take place, only host name will be checked.

https://mongoose.ws/documentation/#struct-mg_tls_opts
> `name` - Server name; an [mg_str](https://mongoose.ws/documentation/#struct-mg_str). If not empty, enable server name verification
